File: console/src/lib/machine.ts
// What the workspace runs on (GET /api/workspace/machine), and the rules for reading it.
//
// The CP answers with two halves that are deliberately not merged
// (control-plane/workspace_machine.go): `measured` is what the container read about itself,
// `declared` is the box this deployment's configuration resolves to. Which one a given row
// ends up showing — and the fact that they can disagree — is decided here rather than in
// each screen: the settings section and the WS-bar popover would otherwise answer the same
// question differently, and the one that drifted would be impossible to spot.

export const MIB = 1048576;
export const GIB = 1073741824;

/** Read from inside the container. A field is absent when it could not be measured — never
 *  zero, so "idle" and "unknown" stay distinguishable (ADR 0058). */
export interface MachineMeasured {
  arch?: string;
  vcpu?: number;
  /** cgroup CPU bandwidth cap in cores. Absent = uncapped, which is the normal case. */
  cpu_quota?: number;
  /** This container's memory limit. */
  mem_max?: number;
  /** The box's RAM. The CP drops it unless the box is this member's own. */
  mem_total?: number;
  disk_total?: number;
  /** Only present where the box could be confirmed to be EC2 (SMBIOS). */
  instance_type?: string;
}

/** The box this deployment's configuration says the workspace gets. Present only on a
 *  runtime that has a box to name (the EC2 slot pool). */
export interface MachineDeclared {
  instance_type?: string;
  arch?: string;
  vcpu?: number;
  slot_mem_mib?: number;
  mem_cap_mib?: number;
  home_gib?: number;
  class_id?: string;
  class_label?: string;
  /** The box belongs to one member, which is what allows its RAM and cores to be shown
   *  as "your machine". */
  dedicated?: boolean;
}

export interface WsMachine {
  runtime: string;
  running: boolean;
  measured?: MachineMeasured;
  declared?: MachineDeclared;
}

/** One row's answer: the value, and whether it was measured or only configured. A row with
 *  neither is not drawn — nothing here invents a number. */
export interface MachineValue<T> {
  value: T;
  measured: boolean;
}

const pick = <T>(m: T | undefined, d: T | undefined, empty: T): MachineValue<T> =>
  m !== undefined && m !== null && m !== empty ? { value: m, measured: true } : { value: d ?? empty, measured: false };

export const machineInstance = (x: WsMachine) => pick(x.measured?.instance_type, x.declared?.instance_type, "");
export const machineArch = (x: WsMachine) => pick(x.measured?.arch, x.declared?.arch, "");
export const machineVCPU = (x: WsMachine) => pick(x.measured?.vcpu, x.declared?.vcpu, 0);

/** The memory the WORKSPACE may spend, which is what a build runs into — not the box's
 *  RAM. The two stopped being the same number once a reserve was held back for the box's
 *  own daemons, and showing the box alone promises memory the cgroup refuses. */
export const machineMemLimit = (x: WsMachine) =>
  pick(x.measured?.mem_max, x.declared?.mem_cap_mib ? x.declared.mem_cap_mib * MIB : 0, 0);

/** The box's own RAM, and only where the box is the member's: on a shared host it is other
 *  people's memory. The CP already withholds the measurement; this keeps the declared side
 *  to the same rule. */
export const machineMemBox = (x: WsMachine) =>
  pick(x.measured?.mem_total, x.declared?.dedicated && x.declared.slot_mem_mib ? x.declared.slot_mem_mib * MIB : 0, 0);

export const machineDisk = (x: WsMachine) =>
  pick(x.measured?.disk_total, x.declared?.home_gib ? x.declared.home_gib * GIB : 0, 0);

/** The box the NEXT start will use, when that is not the one running now — "" when they
 *  agree or when only one of them is known.
 *
 *  A size or class change applies at the next start, so a running workspace keeps its old
 *  box; this is the case someone asking "why is this still slow" is looking at. When the
 *  instance types cannot be compared (no SMBIOS reading), the memory limit answers the
 *  same question: a different rung always means a different cap. */
export function machineNextStart(x: WsMachine): { box: string; changed: boolean } {
  const m = x.measured;
  const d = x.declared;
  if (!m || !d) return { box: "", changed: false };
  if (m.instance_type && d.instance_type) {
    const changed = m.instance_type !== d.instance_type;
    return { box: changed ? d.instance_type : "", changed };
  }
  const cap = d.mem_cap_mib ? d.mem_cap_mib * MIB : 0;
  // A megabyte of slack: the declared cap is a whole number of MiB and the cgroup value is
  // the same number, but nothing guarantees the arithmetic stays exact.
  return { box: "", changed: m.mem_max != null && cap > 0 && Math.abs(m.mem_max - cap) > MIB };
}

/** The one-line form for the WS bar: "m8g.large · arm64 · 2 vCPU", dropping whatever is
 *  unknown. Empty when nothing at all is known, so the caller renders no line rather than
 *  an empty one. */
export function machineSummary(x: WsMachine): string {
  const vcpu = machineVCPU(x).value;
  return [machineInstance(x).value, machineArch(x).value, vcpu ? `${vcpu} vCPU` : ""].filter(Boolean).join(" · ");
}

File: workspace/agent/internal/resources/machine.go
package resources

import (
	"os"
	"runtime"
	"strconv"
	"strings"
)

// Machine is what this container is RUNNING ON, as opposed to Stats, which is how much of
// it is being used right now. Stats rides the CP's 4-second tick; this changes only when
// the container is recreated, so it is its own route and is never merged into that stream.
//
// Same rule as Stats: an axis that cannot be read is OMITTED, never zeroed. "2 vCPU" and
// "could not tell" must not render as the same thing.
//
// What the CP is allowed to forward differs per field, and the Agent does not decide that
// — see MemTotal.
type Machine struct {
	// Arch is the CPU architecture, taken from the compiled binary rather than uname:
	// the Agent ships inside the workspace image, so its GOARCH IS the image's
	// architecture and cannot disagree with what the CLIs on PATH will run as.
	Arch string `json:"arch"`
	// VCPU is how many CPUs this container may run on (the affinity mask), which on a
	// runtime that sets no cpuset is the box's own count. It is a count, not a share:
	// CPUQuota is the share, and the two are independent.
	VCPU int `json:"vcpu,omitempty"`
	// CPUQuota is the cgroup bandwidth cap expressed in cores (1.5 = one and a half).
	// Absent = no quota, which is the case on every runtime the fleet ships today; it
	// exists so a deployment that does cap CPU does not silently show VCPU alone.
	CPUQuota *float64 `json:"cpu_quota,omitempty"`
	// MemMax is this container's own memory limit (cgroup), i.e. what a build may spend.
	MemMax *uint64 `json:"mem_max,omitempty"`
	// MemTotal is the BOX's RAM, read from /proc/meminfo — which is not namespaced, so it
	// is the host's figure whether or not the host belongs to this member.
	//
	// ⚠️ Only meaningful where a workspace has its box to itself (`ecs-ec2`, one slot per
	// member — ADR 0045 decision 8). On docker it is the shared fleet host, and printing
	// it as "your machine" would be someone else's number. The Agent cannot tell the
	// difference, so it reports the fact and the CP decides whether to forward it.
	MemTotal *uint64 `json:"mem_total,omitempty"`
	// DiskTotal is the capacity of the filesystem home sits on (statfs), which on
	// `ecs-ec2` is the persistent EBS volume itself.
	DiskTotal *uint64 `json:"disk_total,omitempty"`
	// InstanceType is the EC2 instance type this container is running on, and is present
	// only when the box could be CONFIRMED to be an EC2 instance (see instanceType).
	InstanceType string `json:"instance_type,omitempty"`
}

// ReadMachine returns what could be established about this container's machine.
func ReadMachine() Machine {
	m := Machine{Arch: runtime.GOARCH, VCPU: runtime.NumCPU()}
	if v, ok := memMax(); ok {
		m.MemMax = &v
	}
	if v, ok := cpuQuotaCores(); ok {
		m.CPUQuota = &v
	}
	if v, ok := hostMemTotal(); ok {
		m.MemTotal = &v
	}
	if _, total, ok := homeUsage(); ok {
		m.DiskTotal = &total
	}
	m.InstanceType = instanceType()
	return m
}

// --- CPU bandwidth ---

// cpuQuotaCores converts the cgroup bandwidth cap into cores. v2's cpu.max is
// "<quota|max> <period>" on one line; v1 splits it into two files and spells "no quota" as
// -1 (which is why it is read as a signed value — parsing it unsigned would produce a
// gigantic core count instead of "unlimited").
func cpuQuotaCores() (float64, bool) {
	if b, err := os.ReadFile(cgroupDir() + "/cpu.max"); err == nil {
		f := strings.Fields(string(b))
		if len(f) == 2 && f[0] != "max" {
			quota, err1 := strconv.ParseFloat(f[0], 64)
			period, err2 := strconv.ParseFloat(f[1], 64)
			if err1 == nil && err2 == nil && period > 0 && quota > 0 {
				return quota / period, true
			}
		}
		return 0, false
	}
	quota, ok1 := readCgroupInt("cpu/cpu.cfs_quota_us")
	period, ok2 := readCgroupInt("cpu/cpu.cfs_period_us")
	if !ok1 || !ok2 || quota <= 0 || period <= 0 {
		return 0, false
	}
	return float64(quota) / float64(period), true
}

func readCgroupInt(name string) (int64, bool) {
	b, err := os.ReadFile(cgroupDir() + "/" + name)
	if err != nil {
		return 0, false
	}
	v, err := strconv.ParseInt(strings.TrimSpace(string(b)), 10, 64)
	return v, err == nil
}

// --- the box ---

func meminfoPath() string {
	if v := os.Getenv("AF_MEMINFO_FILE"); v != "" {
		return v
	}
	return "/proc/meminfo"
}

// hostMemTotal reads MemTotal (in kB) out of /proc/meminfo. Deliberately not the cgroup:
// the cgroup limit is the workspace's share and is already MemMax — this is the box.
func hostMemTotal() (uint64, bool) {
	b, err := os.ReadFile(meminfoPath())
	if err != nil {
		return 0, false
	}
	for _, line := range strings.Split(string(b), "\n") {
		f := strings.Fields(line)
		if len(f) >= 2 && f[0] == "MemTotal:" {
			kb, err := strconv.ParseUint(f[1], 10, 64)
			if err != nil {
				return 0, false
			}
			return kb * 1024, true
		}
	}
	return 0, false
}

func dmiDir() string {
	if v := os.Getenv("AF_DMI_DIR"); v != "" {
		return v
	}
	return "/sys/devices/virtual/dmi/id"
}

// dmiEC2Vendor is what SMBIOS reports for an EC2 instance. It is the whole safety of this
// read: on a Nitro instance product_name holds the instance type ("m8g.large"), but on any
// other machine it holds that machine's model name, and there is nothing in the string
// itself to tell the two apart. Measured on a fleet host: sys_vendor "GMKtec",
// product_name "NucBox G11" — which would have been shown, verbatim, as an instance type.
const dmiEC2Vendor = "Amazon EC2"

// instanceType returns the EC2 instance type, or "" when this is not (provably) EC2.
//
// SMBIOS is read rather than IMDS on purpose: IMDSv2 needs a token round trip and a hop
// limit the deployment does not control, while /sys is already mounted in every container.
//
// ⚠️ Unverified on Graviton and on a real ECS host. Callers must therefore treat "" as
// "not measured", never as "not EC2" — the CP has a declared answer to fall back on.
func instanceType() string {
	if !strings.HasPrefix(dmiRead("sys_vendor"), dmiEC2Vendor) {
		return ""
	}
	t := dmiRead("product_name")
	// An instance type is "<family>.<size>". Anything else means SMBIOS is telling us
	// something we do not understand, and a label nobody can act on is worse than none.
	if !strings.Contains(t, ".") || strings.ContainsAny(t, " /") {
		return ""
	}
	return t
}

func dmiRead(name string) string {
	b, err := os.ReadFile(dmiDir() + "/" + name)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(b))
}
